x-pack/plugins/enterprise_search/public/applications/app_search/components/document_creation/document_creation_logic.ts

```diff
@@ -5,20 +5,27 @@
  */
 
 import { kea, MakeLogicType } from 'kea';
+import dedent from 'dedent';
 
+import { DOCUMENTS_API_JSON_EXAMPLE } from './constants';
 import { DocumentCreationMode, DocumentCreationStep } from './types';
 
 interface DocumentCreationValues {
   isDocumentCreationOpen: boolean;
   creationMode: DocumentCreationMode;
   creationStep: DocumentCreationStep;
+  textInput: string;
+  fileInput: FileList | [];
+  hasFile: boolean; // For some reason fileInput changes don't trigger a rerender, so we can't simply use fileInput.length
```

Comment:
Exactly. React and Redux both require state to be Immutable, so any update to state objects should update the entire reference. Where does this mutation actually occur?

Reply:
Looks like it's at this line below:

```
(files) => files && setFileInput(files)
```

Which means `EuiFilePicker` is the likely culprit.

You could work around this by cloning the array in the handler:
```
(files) => files && setFileInput(files.map(x => x))
```

Reply:
Ahh interesting - I wonder if that's worth it (from a dev readability perspective) to clone vs simply just adding a `hasFiles` reducer? I think the reducer is a little more straightforward to reason about - do you have a preference?

I'd also be a little worried I think about treating `files` as an array vs a `FileList` - FileList is specifically not an array (for some reason) and doesn't necessarily have all the nice array methods like `.map`.

What are your thoughts on storing just a single `file` instead of a `FileList` in that case? So I'd do something like

```
(files) => files && setFileInput([files[0])
```

and change all my type defs from `FileList` to just `File`

Will give that a shot here in a bit 🤔

Reply:
I should add for context also the reason why `files &&` is there in the first place was because Typescript was complaining that `files` was possibly null - not sure where it's getting that info from as that hasn't been my experience really with the filepicker component 🤔

Reply:
https://github.com/elastic/kibana/pull/86056/commits/929724709368b3bb697fa6f5b47378e748953352 - let me know what you think! 😁 Disabled state is looking good on my end, and I tested this on a branch w/ working uploads and it seemed to have no issues there as well.

I actually think I like this a lot more, super happy for the incentive to fix it up.

Reply:
I'll take a look shortly.

FWIW, as a best practice, I believe you should only store simple javascript types (I don't know what the correct word is here) in Redux state. I think the reasoning for that is that you should be able to serialize and deserialize state at any time. Having a type like FileList stored in state is not compatible with that concept.

Reply:
[Just speaking hypothetically now] Ahh interesting - what if (for whatever reason, product-decision wise) we were supporting uploading multiple files instead of just 1 though? We would need to store a FileList then, wouldn't we?

Reply:
In this case you'd need to probably need to store serialize to plain old JS objects before storing it in state, and then de-serialize when retrieving it.

Sounds clunky, but the problem with storing a mutable state object like FileList is that some of the core design features of Redux break. Like if you try to revert to a prior state with Redux Dev Tools, you *will* get prior state, but because FileList was mutated directly, its properties won't revert.

Reply:
ya it's an interesting problem for sure - it's something you run into with plain old objects as well (foo.bar example above), which is something perhaps to keep an eye on